[WIP] Immediate writes to underlying stream until buffer is filled

clue opened the pull request:
With this change applied, the `Buffer` will immediately write to the underlying stream instead of waiting for the loop to say it it writable. This is safe because the stream is non-blocking anyway and the Buffer class will return to normal buffer behavior once the data can not be flushed immediately.

Running the examples/benchmark-throughput.php script, this change alone increases performance from ~100 MiB/s to ~160 MiB/s on my (rather mediocre) test system, using PHP 7.0.8.
Combined with #53, it increases performance from ~100 MiB/s to ~1900 MiB/s.

~~Marking this as WIP because this needs some additional tests once #52 is in.~~ Edit: done.

clue commented:
Afaict this has also been suggested in #15 by @mbonneau around 2 years ago.

WyriHaximus commented:
This combined with #53 puts my benchmark at 3000MB/s where #53 already raised it to 2500MB/s from 260MB/s on my laptop (PHP 7).

clue commented:
Rebased now that #52 is in. Ready for review :shipit:

WyriHaximus commented:
LGTM :shipit:

davidwdan commented:
Can we get this merged?  It bit me today and wasted many hours.

WyriHaximus commented:
Agreed @davidwdan lets get this in. @clue what is holding us back from merging this and related issues?

clue commented on tests/BufferTest.php:
```diff
@@ -129,21 +141,34 @@ public function testWriteDetectsWhenOtherSideIsClosed()
         $buffer = new Buffer($stream, $loop);
         $buffer->softLimit = 4;
         $buffer->on('error', $this->expectCallableNever());
         $buffer->on('drain', $this->expectCallableOnce());
```
Does it really make sense to emit a `drain` event here?

After reconsidering this, we should probably *not* trigger this event if the data can be flushed immediately. Even though the data may exceed the buffer size, this is an atomic step that can not be observed from the outside.

Instead, we should probably only trigger this event if the buffer returned `false` previously and it is *now* ready again.

clue commented:
> Can we get this merged?

Absolutely! :+1: 

> It bit me today and wasted many hours.

Huh, care to elaborate?

clue commented:
Rebased now that #55 is in.

I've moved this PR to the next milestone because it requires an update to not emit a `drain` event during an immediate write. This actually requires quite a few updates in our test suite because most test cases do not expect immediate writes and thus expect a `drain` event.

clue commented:
Removing the milestone for now until the `drain` behavior is addressed.

clue commented:
The main focus is currently on getting the v1.0.0 out. As such, there are currently no immediate plans to build this from my end, so I'm closing this for now :+1:

mbonneau commented:
@clue - Any chance of getting this going again?

clue commented:
@mbonneau I'm currently busy working on the other components and this feature does not seem to have a high priority right now, but in case anybody feels like picking this up, I'm happy to help :shipit: :+1:
